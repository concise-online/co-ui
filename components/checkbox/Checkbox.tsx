import React, { Component, createRef, DOMAttributes, CSSProperties, ChangeEvent, RefObject } from 'react';
import classNames from 'classnames';
import CheckboxGroup from './CheckboxGroup';
import { runCallback } from '../_utils/function';
import { getPrefixCls } from '../_utils/config';

const prefixCls = getPrefixCls('checkbox');

export interface Props extends Omit<DOMAttributes<HTMLInputElement>, 'onChange'> {
  value?: string;
  className?: string;
  checked?: boolean;
  style?: CSSProperties;
  name?: string;
  disabled?: boolean;
  onChange?: (value: string, e: ChangeEvent<HTMLInputElement>) => void;
  forwardedRef?: RefObject<HTMLInputElement>;
}

class Checkbox extends Component<Props> {
  static Group = CheckboxGroup;
  static isCheckbox = true;
  private ref: RefObject<HTMLInputElement>;
  constructor(props: Props) {
    super(props);

    this.ref = createRef();
    this.handleChange = this.handleChange.bind(this);
  }

  focus() {
    (this.ref.current as HTMLInputElement).focus();
  }

  blur() {
    (this.ref.current as HTMLInputElement).blur();
  }

  handleChange(e: ChangeEvent<HTMLInputElement>) {
    const { disabled, onChange } = this.props;

    if (disabled) return;
    runCallback(onChange, e.target.checked, e);
  }

  render() {
    const { children, className, style, disabled, checked, ...checkboxProps } = this.props;
    const classes = classNames(prefixCls, {
      [`${prefixCls}-checked`]: checked,
      [`${prefixCls}-disabled`]: disabled,
    }, className);

    return (
      <label className={classes} style={style}>
        <span className={`${prefixCls}-element`}>
          <input
            {...checkboxProps}
            type="checkbox"
            checked={checked}
            disabled={disabled}
            onChange={this.handleChange}
            ref={this.ref}
          />
        </span>
        <span className={`${prefixCls}-text`}>
          {children}
        </span>
      </label>
    );
  }
}

export default Checkbox;
