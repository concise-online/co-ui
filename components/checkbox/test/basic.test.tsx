import React from 'react';
import { mount, render } from 'enzyme';
import Checkbox from '../index';

describe('checkbox basic feature', () => {
  it('renders correctly', () => {
    const wrapper = render(<Checkbox />);
    const wrapper1 = render(<Checkbox>Checkbox</Checkbox>);

    expect(wrapper).toMatchSnapshot();
    expect(wrapper1).toMatchSnapshot();
  });

  it('change event correctly', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const onChange = jest.fn();
    const wrapper = mount((
      <Checkbox value="foo" onChange={onChange}>
        Checkbox
      </Checkbox>
    ), { attachTo: container });

    wrapper.find('input').simulate('change');
    expect(onChange).toBeCalled();
  });

  it('focus method', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const onFocus = jest.fn();
    const wrapper = mount((
      <Checkbox onFocus={onFocus}>
        Checkbox
      </Checkbox>
    ), { attachTo: container });

    (wrapper.instance() as Checkbox).focus();
    expect(onFocus).toBeCalled();
  });

  it('blur method', () => {
    const container = document.createElement('div');
    document.body.appendChild(container);
    const onBlur = jest.fn();
    const wrapper = mount((
      <Checkbox onBlur={onBlur}>
        Checkbox
      </Checkbox>
    ), { attachTo: container });

    (wrapper.instance() as Checkbox).focus();
    (wrapper.instance() as Checkbox).blur();
    expect(onBlur).toBeCalled();
  });
});
