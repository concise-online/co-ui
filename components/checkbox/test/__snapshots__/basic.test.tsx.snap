// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkbox basic feature renders correctly 1`] = `
<label
  class="co-checkbox"
>
  <span
    class="co-checkbox-element"
  >
    <input
      type="checkbox"
    />
  </span>
  <span
    class="co-checkbox-text"
  />
</label>
`;

exports[`checkbox basic feature renders correctly 2`] = `
<label
  class="co-checkbox"
>
  <span
    class="co-checkbox-element"
  >
    <input
      type="checkbox"
    />
  </span>
  <span
    class="co-checkbox-text"
  >
    Checkbox
  </span>
</label>
`;
